### DISCUSSION TITLE
🙏 Testers wanted 🙏

### DISCUSSION THREAD

**Discussion by hrvach** [hrvach/deskhop#94] [Q&A]
I am working on extending support for some devices that had issues historically and would kindly ask for some help with testing the upcoming releases.

Even if your keyboard/mouse works perfectly, I'd still appreciate your input on an upcoming release to make sure nothing got accidentally broken for you.

If your keyboard/mouse wasn't working, your input is especially appreciated - despite the efforts, it's hard to fix issues without access to the hardware that doesn't work. Your help is much appreciated!

[Update] Newest version:

[deskhop.zip](https://github.com/user-attachments/files/15859217/deskhop.zip)

Thank you!

**Comment by chri2** [comment]
> First item to test - [testing-release.zip](https://github.com/hrvach/deskhop/files/15447183/testing-release.zip)

I'll try as soon as I'll find time. But: could you upload the source as a branch so it'd be possible to compile it with my own settings? With the default settings the mouse for sure is not usable in my setup and testing is quite difficult.

**Reply by hrvach** [reply]
> I'll try as soon as I'll find time. But: could you upload the source as a branch so it'd be possible to compile it with my own settings? With the default settings the mouse for sure is not usable in my setup and testing is quite difficult.

Of course I'll upload the source, just have to tidy it up a bit so it's not totally embarrassing :)

**Comment by raldnor** [comment]
Hi Hrvach,

I just tried this release and the rotary (volume) encoder works, including mute (a press on the encoder), but fn+esc does not work with this firmware. I mapped fn+esc to the power button in the QMK firmware.

Cheers!

**Reply by hrvach** [reply]
> I just tried this release and the rotary (volume) encoder works, including mute (a press on the encoder), but fn+esc does not work with this firmware. I mapped fn+esc to the power button in the QMK firmware.

That's kind of expected, since system control section is next on my to-do list - I'll update here when I get it done. Thanks!

**Comment by jalmeroth** [comment]
Hi @hrvach ,

for me the device is constantly resetting or switching the active board back and forth.

This is what I saw when pressing an 'a' on the keyboard.
![swappy-20240526_161512](https://github.com/hrvach/deskhop/assets/4163868/cecdbe6c-501f-4cdc-903f-d11a2e5ae440)

Here's the full pcap (Wireshark) file for this one session:
[deskhop_testing_release.zip](https://github.com/hrvach/deskhop/files/15447768/deskhop_testing_release.zip)

Hope this helps,
Jan

**Comment by hrvach** [comment]
@jalmeroth apologies, I'm a dumbass and mistakenly uploaded debug version of the firmware. Updated, sorry! :)

**Reply by jalmeroth** [reply]
Hehe, no you're definetely not, no worries. I tried v2 but its still crash-looping for me :shrug:

**Reply by hrvach** [reply]
> Hehe, no you're definetely not, no worries. I tried v2 but its still crash-looping for me 🤷

Darn, I'm confused. Does it happen without kbd/mouse plugged in as well? 🤦 🤦 🤦

**Reply by jalmeroth** [reply]
It's stable w/o adapter, even when the adapter gets connected, but once I hit a key it starts to go wild

**Reply by jalmeroth** [reply]
It's happening after these two interripts at the end, which are not initiated by any action from my side
![swappy-20240526_192316](https://github.com/hrvach/deskhop/assets/4163868/23af225d-729f-4808-8007-f0f64b5ebb2d)

**Reply by hrvach** [reply]
> It's stable w/o adapter, even when the adapter gets connected, but once I hit a key it starts to go wild

I updated tinyusb in this release, maybe something changed? Or I just made a bug I can't find :) 
[testing-release-3.zip](https://github.com/hrvach/deskhop/files/15448362/testing-release-3.zip)
This one is built with the previous (current prod) version of tinyusb. Same thing?

**Reply by hrvach** [reply]
> It's stable w/o adapter, even when the adapter gets connected, but once I hit a key it starts to go wild

Also can you please describe your keyboard/receiver usb vid/pid? I need to get one :)

**Reply by jalmeroth** [reply]
> Also can you please describe your keyboard/receiver usb vid/pid? I need to get one :)

All of this info has been collected here: https://github.com/hrvach/deskhop/issues/47.
I will test the new release later and report back

**Reply by hrvach** [reply]
> > Also can you please describe your keyboard/receiver usb vid/pid? I need to get one :)
> 
> All of this info has been collected here: #47. I will test the new release later and report back

Thank you so much! I've found at least one bug in the meantime (despite it working for me with no issues), please hold on while I fix it :)

**Reply by jalmeroth** [reply]
v4 is stable for me now :+1: But the parsed/sent keyboard values are not sensible yet.

**Comment by hrvach** [comment]
This one should be better.

[testing-release-4.zip](https://github.com/hrvach/deskhop/files/15455062/testing-release-4.zip)

**Reply by chris-010** [reply]
no problem so far

**Reply by Faisalalali** [reply]
Hello, new tester here.
I'm noticing that mouse movement is inconsistent on both Windows and Linux. Kinda like if the acceleration is skewed to the bottom right somehow. I tested a typical mouse and a Logetich MX mouse and they both show same behavior.
Has the mouse movements emulation changed in this version?

**Reply by hrvach** [reply]
> Hello, new tester here. I'm noticing that mouse movement is inconsistent on both Windows and Linux. Kinda like if the acceleration is skewed to the bottom right somehow. I tested a typical mouse and a Logetich MX mouse and they both show same behavior. Has the mouse movements emulation changed in this version?

Well caught - respect, it's a regression of a previously fixed bug. And embarrassing enough, I've been using it without noticing :)

**Comment by hrvach** [comment]
[New version](https://github.com/user-attachments/files/15859217/deskhop.zip) with LOTS of changes is up for testing while I tidy and commit the code. Most keyboard shortcuts related to configuration have been removed in favor of a web GUI.

Changes and updates:

1. UART reworked to use DMA without the need for CPU to worry about serial comms.

2. Single unified firmware, just one .uf2 which should hopefully autodetect where it's running.

3. One box should be able to update the other if there is a higher version detected running. You will need to only update one of. them, the other will copy its firmware and update itself.

4. Firmware upgrade can now be done easier, with board presenting itself as a usb stick while still running, so you can use mouse/keyboard and not have to plug them directly - everyone hates tracking cables.

5. There is a new GUI "configuration mode" activated by "left shift + right shift + C + O". This causes the box on the kbd side to reboot and present itself with additional configuration HID endpoint and a mass storage device (like a USB drive). On that drive, you'll find a "config.htm" file you can open using Chromium/Chrome - sadly Firefox has no webhid :-(( This presents a web form to configure the device in an easy way, without having to remember 400 weird shortcuts. Hit connect first, device should be listed (but only if it's in configuration mode).

<img width="1889" alt="Screenshot 2024-06-16 at 19 34 15" src="https://github.com/hrvach/deskhop/assets/7192749/1727733b-d608-457c-a2f3-e4487727713f">

6. Linker script changed, build will have a new script that calculates crc32 and updates the binary so the box can verify flashing process validity.

7. HID parser reworked, some nkro keyboards should be supported.

I probably forgot something. Thanks for any feedback and testing...

**Reply by chri2** [reply]
**Nice!**

* installed on both Picos fine
* changed to config mode when hotkey pressed
* opened config.htm on a Debian Bookworm in chromium and ungoogled-chrome
* when hitting connect I can select the Deskhop ("Paired")

what does not work in either of the two browsers:

* blink
* blink both
* could not save and later read settings, settings didn't have an effect
* tried various combinations of connect followed by blink/read/save/wipe config - no effect
* never got the fields in the Device Status sections filled

improvements / suggestions:

* config.htm presents "Output A | Output B" and on the board I read "B (mouse) | A (mouse)" - did the board layout change in the new board version? I'd prefer to have config.htm and the board layout show the same order.
* couldn't test the Speed settings, because they didn't work. I expect that the freedom to use numbers allowed a much finer tuning. Same for Jump Treshold.
* Hotkey Toggle is missing the 'Pause' key

Most of the testing I did in Chromium Version 125.0.6422.141.

**Reply by hrvach** [reply]
Thanks for trying it out, that's why testing is important. My assumption would be permissions - you might need an udev rule in /etc/udev/rules.d/deskhop.rules:

KERNEL=="hidraw*", SUBSYSTEM=="hidraw", ATTRS{idVendor}=="1209", ATTRS{idProduct}=="c000", MODE="0660", GROUP="wheel"

use "id" to check the groups your user is a member of and pick one, but usually going with your username is a good bet :) 

udevadm control -R

then reconnect and re-try?

**Reply by jalmeroth** [reply]
Hi @hrvach,

thanks for this massively revamped version! The changelog reads awesome.

Unfortunately, I'm getting more reluctant to test these new versions, while there's no code released. I know you want to polish the code first, but the changes in this release are quite big. I wouldn't know where to even start testing it. Instead, I'm going to quote the "Security and Safety" section from the README here:

> Entirety of the code is open source, without any binary blobs and thoroughly commented to explain its purpose. I encourage you to never trust anyone and always make sure you know what you are running by doing a manual audit.

Please consider a development branch or something similiar so we can provide more directed feedback/testing.

Thanks for all the efforts you put into this project. :rocket:

**Reply by hrvach** [reply]
@jalmeroth that's totally understandable, I know it's a lot to ask. I'm updating the code/docs and code should be available shortly - hoping for some feedback then. Cheers!

**Reply by chri2** [reply]
The udev rule helped. Most settings seem to work. The settings for speed I liked as numbers a lot better. I've got the feeling I'm somewhere inbetween now.

I couldn't find any effect when changing 'Jump Threshold' and 'Hostkey Toggle'.

Border Top and Border Bottom need some more explaining.

**Reply by hrvach** [reply]
> The udev rule helped. Most settings seem to work. The settings for speed I liked as numbers a lot better. I've got the feeling I'm somewhere inbetween now.
> 
> I couldn't find any effect when changing 'Jump Threshold' and 'Hostkey Toggle'.
> 
> Border Top and Border Bottom need some more explaining.

Thank you for the feedback, that's exactly what I need - very often what I assume would be OK might not seem logical to others, so this is very useful feedback.

I may or may not admit I forgot to "wire in" those two settings. 🤣 

Border top/bottom are calibration values, when you have two mismatched screens and you want the cursor to come out at the same height. I have to figure out a nice, intuitive way to calibrate that. So overall, the most important question - in terms of keyboard/mouse, it's not worse than the stable release?

**Reply by chri2** [reply]
> it's not worse than the stable release?

Hm, I used it all day and the only thing somehow worth is that I can't adjust the Speed parameters finer. But beside that it works.

Ah, maybe one more thing I didn't look into: I could'nt get used to the INS and POS1 key position on my actual keyboard and used udev hwdb settings to change there meaning. This doesn't work anymore. I'll look into that later, but I guess it'll just be changed USB vendor:product numbers...

**Reply by hrvach** [reply]
> > it's not worse than the stable release?
> 
> Hm, I used it all day and the only thing somehow worth is that I can't adjust the Speed parameters finer. But beside that it works.
> 
> Ah, maybe one more thing I didn't look into: I could'nt get used to the INS and POS1 key position on my actual keyboard and used udev hwdb settings to change there meaning. This doesn't work anymore. I'll look into that later, but I guess it'll just be changed USB vendor:product numbers...

The speed I'll fix, thanks for noticing! The vid:pid are different only while in configure mode, which is not supposed to be left running (for security reasons). I'll consider adding a timeout and a flashing light or something... :)

**Reply by chri2** [reply]
> The vid:pid are different only while in configure mode, which is not supposed to be left running (for security reasons). I'll consider adding a timeout and a flashing light or something... :)

That's a good idea. I just closed the browser and that's why it still had been on. I found the `exit` button now and the keyboard keys are back at their designated places ;-)  - thanks a lot for the explanation!

**Reply by hrvach** [reply]
> > The vid:pid are different only while in configure mode, which is not supposed to be left running (for security reasons). I'll consider adding a timeout and a flashing light or something... :)
> 
> That's a good idea. I just closed the browser and that's why it still had been on. I found the `exit` button now and the keyboard keys are back at their designated places ;-) - thanks a lot for the explanation!

Thank you, you've managed to suggest 5 things I missed and 3 things to improve. You're the new head of quality assurance! :)

**Reply by chri2** [reply]
Haha, you know the saying ["Die Geister, die ich rief…"](https://en.wikipedia.org/wiki/The_Sorcerer%27s_Apprentice#German_culture)

Your project is amazing and I follow it with great joy! Thanks a lot!

BTW: A **question about security** came to my mind: I really liked the idea that the device could not be compromised without the need of touching the hardware (well, yes, there's already been that hotkey to switch to upgrade mode…). The setup becomes by the automatic update between the two picos less secure.

For changes to the software or configuration I'd love to have a security mode that needs the user to (there do not seem to be many options) push the button to confirm actions (or if there are options, short pins, touch something - whatever makes it necessary to actually have access to the hardware to confirm actions).

A nice example for this kind of added security is the Nitrokey3: it blinks and waits for a touch confirmation if configured to do so.

**Reply by chris-010** [reply]
Here are my findings:

Windows 10
Chrome Version 126.0.6478.62

- installed test fw on both nodes
- changed to config mode when hotkey pressed
- Problem1: CAPSLOCK stays active, can just be disabled by pressing capslock on keyboard connected to deskhop
- Problem2: After disabling capslock, new strage behavior, it seems that shift is still active. You can't see this on keyboard input, but on mouse input, cause text is getting marked from outgoing from the coordinates where Problem1 appears.
- Workaround: CTRL+ALT+DEL -> ESC
- opened config.htm on a Windows 10 with Chrome
- so far so good, but not able to connect

![image](https://github.com/hrvach/deskhop/assets/10660568/3c033f6b-b30f-4e63-9391-d2fd6139ed2d)

Maybe this will help:
chrome://device-log/

USBError[15:57:20] SetupDiEnumDeviceInterfaces: Es sind keine Daten mehr verfügbar. (0x103)

USBError[15:57:20] SetupDiGetDeviceProperty({{A45C254E-DF1C-4EFD-8020-67D146A850E0}, 6}) failed: Element nicht gefunden. (0x490)

HIDUser[15:57:20] HID device added: vendorId=4617, productId=49152, name='DeskHop Switch', serial='0', deviceIds=['\\?\hid#vid_1209&pid_c000&mi_00&col03#7&3d487ef&0&0002#{4d1e55b2-f16f-11cf-88cb-001111000030}'], reportDescriptor=''

USBUser[15:57:19] USB device added: path=\\?\usb#vid_1209&pid_c000#0#{a5dcbf10-6530-11d2-901f-00c04fb951ed} vendor=4617 "Hrvoje Cavrak", product=49152 "DeskHop Switch", serial="0", driver="usbccgp", guid=a921133d-6a08-4d67-99ad-6a01416b0f4d

USBUser[15:57:19] USB device removed: path=\\?\usb#vid_2e8a&pid_107c#0#{a5dcbf10-6530-11d2-901f-00c04fb951ed} guid=6d6cce65-23dd-41fe-929d-7fd5622c3895

HIDUser[15:57:19] HID device removed: deviceId='\\?\hid#vid_2e8a&pid_107c&mi_01#7&1da440a4&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}'

HIDUser[15:57:19] HID device removed: deviceId='\\?\hid#vid_2e8a&pid_107c&mi_00&col01#7&3567cce9&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}\kbd'

USBError[15:57:04] SetupDiGetDeviceProperty({{A45C254E-DF1C-4EFD-8020-67D146A850E0}, 6}) failed: Element nicht gefunden. (0x490)

USBError[15:57:03] SetupDiEnumDeviceInterfaces: Es sind keine Daten mehr verfügbar. (0x103)

HIDUser[15:57:03] HID device added: vendorId=11914, productId=4220, name='DeskHop Helper', serial='0', deviceIds=['\\?\hid#vid_2e8a&pid_107c&mi_01#7&1da440a4&0&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}'], reportDescriptor=''

USBError[15:57:03] SetupDiEnumDeviceInterfaces: Es sind keine Daten mehr verfügbar. (0x103)

USBError[15:57:03] SetupDiGetDeviceProperty({{A45C254E-DF1C-4EFD-8020-67D146A850E0}, 6}) failed: Element nicht gefunden. (0x490)

HIDUser[15:57:03] HID device added: vendorId=11914, productId=4220, name='DeskHop Switch', serial='0', deviceIds=['\\?\hid#vid_2e8a&pid_107c&mi_00&col03#7&3567cce9&0&0002#{4d1e55b2-f16f-11cf-88cb-001111000030}'], reportDescriptor=''

USBUser[15:57:03] USB device added: path=\\?\usb#vid_2e8a&pid_107c#0#{a5dcbf10-6530-11d2-901f-00c04fb951ed} vendor=11914 "Hrvoje Cavrak", product=4220 "DeskHop Switch", serial="0", driver="usbccgp", guid=6d6cce65-23dd-41fe-929d-7fd5622c3895

**Reply by hrvach** [reply]
> Haha, you know the saying ["Die Geister, die ich rief…"](https://en.wikipedia.org/wiki/The_Sorcerer%27s_Apprentice#German_culture)

Ich habe das heute gelernt, vielen Dank :)

> Your project is amazing and I follow it with great joy! Thanks a lot!
> 
> BTW: A **question about security** came to my mind: I really liked the idea that the device could not be compromised without the need of touching the hardware (well, yes, there's already been that hotkey to switch to upgrade mode…). The setup becomes by the automatic update between the two picos less secure.
> 
> For changes to the software or configuration I'd love to have a security mode that needs the user to (there do not seem to be many options) push the button to confirm actions (or if there are options, short pins, touch something - whatever makes it necessary to actually have access to the hardware to confirm actions).
> 
> A nice example for this kind of added security is the Nitrokey3: it blinks and waits for a touch confirmation if configured to do so.

That is indeed a good idea, it's never easy to strike a good balance between security and ease of use. The current model is relying on the user explicitly expressing consent to upgrade by enabling the upgrade mode via a 4-button shortcut. This will have an added layer of security - flashing lights to indicate config mode, a 5 minute timer to automatically exit config mode if the user forgets it and only one board is in config mode.

When config mode is off, the config hid / msc do not exist and the only information allowed to flow in is 1 byte of keyboard LED information (might add an option to fully disable this).

Do you think it would make sense to require additionally pushing the button on the PICO for FW upgrade? Having to open the case is annoying to the user and the effect would be the same.

**Reply by chri2** [reply]
> Do you think it would make sense to require additionally pushing the button on the PICO for FW upgrade? Having to open the case is annoying to the user and the effect would be the same.

attack vector:

* user uses a programmable keyboard and there is a way to program it even with deskhop inbetween (no idea if that is the truth or will be possible in the future)
* attacker did compromise computer A
* attacker places a program to watch over the status of deskhop on computer A
* attacker programs the keyboard to send the key combination to enable programming mode of deskhop
* when deskhop switches to programable mode the watchdog on computer A briefly flashes a firmware that alters deskhop in a way that suites the attacker (could be for example to open a channel to make keyboard/mouse of computer B secretly available to computer A

This works also with some hidden malware just watching the state of deskhop and acting after finding it entering programming mode - even though this might not be as successful, because the user would have a chance to notify that something _is different_ from other times they used that mode.

In my opinion for trust into the security of deskhop it is the most important thing to make it secure. If a (or better **the**) feature of deskhop is real separation of the two connected computers than there is no compromise to usability or people working in environments relying on this separation will have to switch back to mechanical kvm switches 😉.

To give an example:

Computer A: Android Phone
Computer B: Desktop 

Computer A needs to be untrusted: it has usually its own connection to some networks (internet, mobile, wlan) while Computer B is in a security domain with rules enforced by software and firewalls.

In this scenario it is not possible to be compliant and use deskhop if Computer A does have the slightest chance to break the deskhop barrier by programming it.

**Reply by hrvach** [reply]
>     * user uses a programmable keyboard and there is a way to program it even with deskhop inbetween (no idea if that is the truth or will be possible in the future)

Programming or configuring QMK-like keyboards isn't supported on purpose, for exactly this reason - however impractical at times.
 
>     * attacker programs the keyboard to send the key combination to enable programming mode of deskhop

If the attacker is able to program your keyboard, it's already game over. Would it make sense to have an "extra safe" option you could enable to enforce options like "press button to allow upgrade" or even "permanently disable config mode"?

**Reply by chri2** [reply]
> If the attacker is able to program your keyboard, it's already game over. 

In my example on Computer A - the one that is not trustworthy. But how to get to Computer B if deskhop cannot be compromised?

In my opinion this is all about separating the two computers sharing keyboard/mouse as much as possible and I would enable every layer of security I could get to ensure this.

**Reply by chris-010** [reply]
I follow this discussion about the consideration between security and usability with interests. The main topic is an assumed security issue that could make it possible to modify firmware without physical user interaction to take over an isolated system. I consider viewpoints from both sides. Therefore i suggest an idea to allow firmware updates only with physical user interaction (to keep security), but without opening the case (to keep usability).

What about to make a two-stage firmware update process.

- **Step1** (w/o physical user interaction): to select firmware file and prepare board for Step2
- **Step2** (with physical user interaction): to flash an finalyze firmware update. But physical interaction is not activated by pressing a button of  rpi pico which is inside a case. Step2 can only be actived by unplug/replug or similar procedure of connected hid device(s).

In short, using the usb plug for physical user interaction to release firmware update process.

**Reply by chri2** [reply]
> using the usb plug for physical user interaction to release firmware update process

Nice 👍. A quite simple way has been used by eletronic devices with upgradeable firmware for a long time: put firmware on usb memory with a certain name, put usb memory into device and wait.

This would make all of this a already known and simple procedure for the user. If it is possible in hardware/software I can't say.

**Reply by chris-010** [reply]
great input, maybe i was thinking too complicated.

**Comment by alankeny** [comment]
I'm running the test firmware from this thread that was built on June 16th.  I have a Logitech Unifying receiver, MX keys keyboard, and M525 mouse.  The Unifying receiver is connected to the keyboard port.

For the most part everything with the mouse and keyboard works as expected except the media keys on the keyboard including mute, volume up, and volume down.  Let me know if there's any data I can collect or troubleshooting steps I should take.  Thanks!

**Reply by hrvach** [reply]
> I'm running the test firmware from this thread that was built on June 16th. I have a Logitech Unifying receiver, MX keys keyboard, and M525 mouse. The Unifying receiver is connected to the keyboard port.
> 
> For the most part everything with the mouse and keyboard works as expected except the media keys on the keyboard including mute, volume up, and volume down. Let me know if there's any data I can collect or troubleshooting steps I should take. Thanks!

I've been trying to get better support for consumer control "variable" format and non-boot mode keyboards, any improvement with [this one](https://github.com/user-attachments/files/15944343/deskhop.zip)?

@chris-010 this one should also fix the windows connectivity issue for configuration mode.

Now for the code cleanup ... :-|

**Reply by alankeny** [reply]
[This one](https://github.com/user-attachments/files/15944343/deskhop.zip) and v0.60 appear to be the same version.  Let me know if I got that wrong, and I'll manually upgrade both picos to v0.60.

The keyboard behavior is the same with this release as the one I started with built on June 16th.  The media keys for mute, volume down, and volume up result in mouse movement and sometimes right-clicks rather than the media keys registering on the host.  Let me know if you need me to do anything.  Thanks!

**Reply by hrvach** [reply]
> [This one](https://github.com/user-attachments/files/15944343/deskhop.zip) and v0.60 appear to be the same version. Let me know if I got that wrong, and I'll manually upgrade both picos to v0.60.
> 
> The keyboard behavior is the same with this release as the one I started with built on June 16th. The media keys for mute, volume down, and volume up result in mouse movement and sometimes right-clicks rather than the media keys registering on the host. Let me know if you need me to do anything. Thanks!

Yes, it's the same version. Oh man, if I knew how many variations this HID thing could have, I'd do a different kind of project :) Perhaps dumping the keyboard hid descriptor would be helpful to figure out what it's parsing and why is it misbehaving. Thanks for your patience :)

**Reply by alankeny** [reply]
First, thanks for creating this!  It's already so much more useful than the Synergy / Barrier / Input Leap software-only solution, because it doesn't depend on the local network to function.  I can use it with my work computer, even while I'm connected to the work VPN which blocks local network access.

Hopefully I've collected what you're asking for.  I ran `lsusb` and got:

    Bus 005 Device 038: ID 046d:c52b Logitech, Inc. Unifying Receiver

I ran `usbhid-dump -a 5:38` and got:

    005:038:002:DESCRIPTOR         1719166459.222426
     06 00 FF 09 01 A1 01 85 10 95 06 75 08 15 00 26
     FF 00 09 01 81 00 09 01 91 00 C0 06 00 FF 09 02
     A1 01 85 11 95 13 75 08 15 00 26 FF 00 09 02 81
     00 09 02 91 00 C0 06 00 FF 09 04 A1 01 85 20 95
     0E 75 08 15 00 26 FF 00 09 41 81 00 09 41 91 00
     85 21 95 1F 09 42 81 00 09 42 91 00 C0

    005:038:001:DESCRIPTOR         1719166459.226667
     05 01 09 02 A1 01 85 02 09 01 A1 00 95 10 75 01
     15 00 25 01 05 09 19 01 29 10 81 02 95 02 75 0C
     16 01 F8 26 FF 07 05 01 09 30 09 31 81 06 95 01
     75 08 15 81 25 7F 09 38 81 06 95 01 05 0C 0A 38
     02 81 06 C0 C0 05 0C 09 01 A1 01 85 03 95 02 75
     10 15 01 26 FF 02 19 01 2A FF 02 81 00 C0 05 01
     09 80 A1 01 85 04 95 01 75 02 15 01 25 03 09 82
     09 81 09 83 81 00 75 06 81 03 C0 06 BC FF 09 88
     A1 01 85 08 95 01 75 08 15 01 26 FF 00 19 01 29
     FF 81 00 C0

    005:038:000:DESCRIPTOR         1719166459.229314
     05 01 09 06 A1 01 95 08 75 01 15 00 25 01 05 07
     19 E0 29 E7 81 02 81 03 95 05 05 08 19 01 29 05
     91 02 95 01 75 03 91 01 95 06 75 08 15 00 26 FF
     00 05 07 19 00 2A FF 00 81 00 C0

**Reply by hrvach** [reply]
> First, thanks for creating this! It's already so much more useful than the Synergy / Barrier / Input Leap software-only solution, because it doesn't depend on the local network to function. I can use it with my work computer, even while I'm connected to the work VPN which blocks local network access.
> 
> Hopefully I've collected what you're asking for. I ran `lsusb` and got:
> 
> ```
> Bus 005 Device 038: ID 046d:c52b Logitech, Inc. Unifying Receiver
> ```
> 
> I ran `usbhid-dump -a 5:38` and got:
> 
> ```
> 005:038:002:DESCRIPTOR         1719166459.222426
>  06 00 FF 09 01 A1 01 85 10 95 06 75 08 15 00 26
>  FF 00 09 01 81 00 09 01 91 00 C0 06 00 FF 09 02
>  A1 01 85 11 95 13 75 08 15 00 26 FF 00 09 02 81
>  00 09 02 91 00 C0 06 00 FF 09 04 A1 01 85 20 95
>  0E 75 08 15 00 26 FF 00 09 41 81 00 09 41 91 00
>  85 21 95 1F 09 42 81 00 09 42 91 00 C0
> 
> 005:038:001:DESCRIPTOR         1719166459.226667
>  05 01 09 02 A1 01 85 02 09 01 A1 00 95 10 75 01
>  15 00 25 01 05 09 19 01 29 10 81 02 95 02 75 0C
>  16 01 F8 26 FF 07 05 01 09 30 09 31 81 06 95 01
>  75 08 15 81 25 7F 09 38 81 06 95 01 05 0C 0A 38
>  02 81 06 C0 C0 05 0C 09 01 A1 01 85 03 95 02 75
>  10 15 01 26 FF 02 19 01 2A FF 02 81 00 C0 05 01
>  09 80 A1 01 85 04 95 01 75 02 15 01 25 03 09 82
>  09 81 09 83 81 00 75 06 81 03 C0 06 BC FF 09 88
>  A1 01 85 08 95 01 75 08 15 01 26 FF 00 19 01 29
>  FF 81 00 C0
> 
> 005:038:000:DESCRIPTOR         1719166459.229314
>  05 01 09 06 A1 01 95 08 75 01 15 00 25 01 05 07
>  19 E0 29 E7 81 02 81 03 95 05 05 08 19 01 29 05
>  91 02 95 01 75 03 91 01 95 06 75 08 15 00 26 FF
>  00 05 07 19 00 2A FF 00 81 00 C0
> ```

This looks more or less identical to my keyboard (the consumer control), but I need to get one of those unifying receivers, seems these are pretty popular and might behave differently. What's your keyboard/mouse model?

**Comment by alankeny** [comment]
> What's your keyboard/mouse model?

My primary setup has a Logitech MX Keys Advanced Wireless Illuminated Keyboard  and a Logitech Wireless Mouse M525.

I also tried a different Unifying receiver that's configured for a Logitech Wireless Touch Keyboard K400.  The K400 includes a keyboard and a touchpad with some dedicated volume keys.  I got similar results on both.

**Reply by hrvach** [reply]
> > What's your keyboard/mouse model?
> 
> My primary setup has a Logitech MX Keys Advanced Wireless Illuminated Keyboard and a Logitech Wireless Mouse M525.
> 
> I also tried a different Unifying receiver that's configured for a Logitech Wireless Touch Keyboard K400. The K400 includes a keyboard and a touchpad with some dedicated volume keys. I got similar results on both.

Could you please try [deskhop-test2.zip](https://github.com/user-attachments/files/15957042/deskhop-test2.zip) and let me know if this fixes it? I have an idea what might be happening here.

**Comment by alankeny** [comment]
All of my media keys are working with [deskhop-test2.zip](https://github.com/user-attachments/files/15957042/deskhop-test2.zip) 👍

Just to confirm I've tested show workspaces, show notifications, show desktop, previous track, play/pause, next track, mute, volume down, and volume up.  They're all working with the Logitech MX Keys.  Thanks so much!

**Comment by ak66666** [comment]
Thank you for this update - the mouse key  and touchpad from the keyboard are recognized now.
I'm running the test firmware from this thread that was built on June 16th.

Mouse: Logitec M545 with unified receiver.
Keyboard: QMK-based ThumbsUp with NKRO and mouse keys enabled, and with Cirque Proglide touchpad.

Mouse works as expected.
Regular keys on the keyboard work as expected too.

Keyboard's mouse key and touchpad seem to be misread/mapped to different direction. 
The QMK left and right mouse directional keys and the touchpad left and right gestures result in the mouse pointer movement up and down.
The up and down mouse key/touchpad seem to be treated as mouse scroll up/down,

**Reply by ak66666** [reply]
I managed to make it work, may be a crude solution, but it works for me.
Submitted a PR: ak66666:rc_ak66666

**Comment by hrvach** [comment]
Is there anyone with multiple monitors on windows who could test something for me? I have an idea how to support multiple screens without it being shitty (as it currently is), but have nothing to test on. Any help appreciated. Thanks!

**Reply by ak66666** [reply]
Hi,
Sure, I can test it, got two Windows machines with 2 and 3 displays.

**Reply by Trekky12** [reply]
Hey,

I have two windows machines: the first one is a notebook with a external display connected and the second one has three displays.
I could do some tests.

**Comment by hrvach** [comment]
@ak66666  Can you check  [test-digitizer.zip](https://github.com/user-attachments/files/16249794/test-digitizer.zip) ? It has (temporarily) absolute mouse replaced by a digitizer pen, expect it to be pretty broken for pretty much everything but I'm just curious how does the cursor behave on multi-screen windows (with number of screens set to 1 / settings factory reset). I've found a microsoft note that gives some clues about the possibility of an actual workaround..

**Reply by ak66666** [reply]
Mouse moves between the screens as expected.
Mouse buttons don't work though.

**Reply by hrvach** [reply]
> Mouse moves between the screens as expected. Mouse buttons don't work though.

That's expected because buttons are not implemented (I know), it's just a test .... but, mouse moves across multiple desktops and it behaves normally? No shitty behaviour like the current prod/rc version have with multiple screen Windows?

**Reply by ak66666** [reply]
Yes, it moves as expected, even the QMK mouse keys are working.
My screens are located a bit weirdly, physically they are set like this:
_______________B3
B2__B1__A1    
_______________A2

In Windows the B-side is like that:

_____B3
____________B1
___B2   

Actually the dev version you published in this thread was working good too - after I set the number of screens and OS type through the new web interface. Without that config there were some issues, but I don't remember how it was behaving then.

This new test version works fine with that layout right off the box, no configuration was needed.

**Reply by hrvach** [reply]
> Yes, it moves as expected, even the QMK mouse keys are working. My screens are located a bit weirdly, physically they are set like this: _______________B3 B2__B1__A1 _______________A2
> 
> In Windows the B-side is like that:
> 
> _____B3 ____________B1 ___B2
> 
> Actually the dev version you published in this thread was working good too - after I set the number of screens and OS type through the new web interface. Without that config there were some issues, but I don't remember how it was behaving then.
> 
> This new test version works fine with that layout right off the box, no configuration was needed.

This is extremely helpful, thank you. I think I'll be able to fix Windows and possibly Mac OS issues with multiple desktops in a way that's not super annoying like now.

**Comment by hrvach** [comment]
I'd appreciate some feedback on v0.72 (pre-release), particularly focusing on NKRO. Works for folks?

**Reply by ak66666** [reply]
Works good for me, no new issues in functionality.

The configuration shows version v.073 though, not 72.

**Reply by hrvach** [reply]
> Works good for me, no new issues in functionality.
> 
> The configuration shows version v.073 though, not 72.

I might be dyslexic, that's probably why :) ... will fix :)

**Comment by alankeny** [comment]
I synced my fork with v.073 / 72 and rebuilt deskhop.uf2.

I configure my Deskhop from a laptop running Linux.  After the update I had to change my udev rules from:
`ATTRS{idVendor}=="1209", ATTRS{idProduct}=="c000"`
to 
`ATTRS{idVendor}=="2e8a", ATTRS{idProduct}=="107c"`

After that change, I could connect, read, and write the web configuration page from Chrome again.  No other issues in functionality.  Everything is still working great.  Thanks!

**Comment by wences-dc-uba-ar** [comment]
Thanks for this awesome gadget, and sorry if this comment not belongs here: 
Today doing update to 0.73 I realized my version was in 175 (don't know since when), but after copying the firmare -> entering_config_again... I noticed the version still showing 175.
After numerous try and error found that have to wipe firmware like the first time install on both picos (while the other was unplugged) to make the update.
(IdK, something about version compare, maybe).
I think a note on the release page could advice on how to ease the update in this numeration-change scenario.

**Comment by hrvach** [comment]
There is a pre-release of v0.77, testers more than welcome! :)

**Reply by ak66666** [reply]
Used it all day today, two windows machines with multiple screens each, no new issues.

My previous issue with mouse/trackball controllers on both input ports still there, yet seem be less frequent than before.
Cursor is still jumping when both side mice are actively used, switched from one to another.
Stops this behavior when one of the trackballs is inactive. 
Yet I'd say it  happens less often than before, and primarily after I switch to the other side mouse like every other minute.
Overall, an improvement.
